core/storage/statistics/gae_models.py

```diff
@@ -587,114 +590,219 @@ def get_versions(cls, exploration_id):
-    try:
-        answer_log.put()
-    except Exception as e:
-        logging.error(e)
-
-
-def resolve_answers(
-        exploration_id, state_name, rule_str, answers):
-    """Resolves selected answers for the given rule.
-
-    Args:
-        exploration_id: the exploration id
-        state_name: the state name
-        rule_str: a string representation of the rule
-        answers: a list of HTML string representations of the resolved answers
-    """
-    assert isinstance(answers, list)
-    answer_log = StateRuleAnswerLogModel.get_or_create(
-        exploration_id, state_name, rule_str)
-
-    for answer in answers:
-        if answer not in answer_log.answers:
-            logging.error(
-                'Answer %s not found in answer log for rule %s of exploration '
-                '%s, state %s, handler %s' % (
-                    answer, rule_str, exploration_id, state_name,
-                    _OLD_SUBMIT_HANDLER_NAME))
-        else:
-            del answer_log.answers[answer]
-    answer_log.put()
-
-
 class StateAnswersModel(base_models.BaseModel):
-    """Store all answers of a state.
+    """Store all answers of a state. This model encapsulates a sharded storage
+    system for answers. Multiple entries in the model may contain answers for
+    the same state. The initial entry has a shard ID of 0 and contains
+    information about how many shards exist for this state. All other meta
+    information is duplicated across all shards, since they are immutable or are
+    local to that shard.
 
     The id/key of instances of this class has the form
-        [EXPLORATION_ID]:[EXPLORATION_VERSION]:[STATE_NAME].
+        [EXPLORATION_ID]:[EXPLORATION_VERSION]:[STATE_NAME]:[SHARD_ID].
     """
+    # This provides about 48k of padding for the other properties and entity
+    # storage overhead (since the max entity size is 1MB).
+    _MAX_ANSWER_LIST_BYTE_SIZE = 100000
+
     # Explicitly store exploration id, exploration version and state name
     # so we can easily do queries on them.
     exploration_id = ndb.StringProperty(indexed=True, required=True)
     exploration_version = ndb.IntegerProperty(indexed=True, required=True)
     state_name = ndb.StringProperty(indexed=True, required=True)
+    shard_id = ndb.IntegerProperty(indexed=True, required=True)
     # Store interaction type to know which calculations should be performed
     interaction_id = ndb.StringProperty(indexed=True, required=True)
+    # Store how many extra shards are associated with this state. This is only
+    # present when shard_id is 0. This starts at 0 (the main shard is not
+    # counted).
+    shard_count = ndb.IntegerProperty(indexed=True, required=False)
+    # The total number of bytes needed to store all of the answers in the
+    # submitted_answer_list, minus any overhead of the property itself. This
+    # value is found by summing all json_size values for answer dicts inside
```

Comment:
The json_size value is an approximate, but it's always an overestimate (which is safe, since we never want to accidentally overflow the entity and fail the transaction).